kernel/irq.c

```diff
@@ -0,0 +1,29 @@
+// SPDX-FileCopyrightText: CGL-KFS
+// SPDX-License-Identifier: BSD-3-Clause
+
+/* irq.c
+ *
+ * Definition of IRQ (Interrupt Request) to handle non-software interrupts
+ * such as keyboard.
+ *
+ * updated: 2022/10/20 - mrxx0 <chcoutur@student.42.fr>
+ * updated: 2022/10/20 - mrxx0 <chcoutur@student.42.fr>
+ */
+
+#include <kernel/pic_8259.h>
+#include <kernel/print.h>
+
+#include "idt_internal.h"
+
+/* KEYBOARD HANDLER (33)
+ * Occurs when 8259 PIC receive an IRQ1 (keyboard hardware interrupt)
+ *
+ * @arg(int_frame): interrupt frame structure.
+ */
+__attribute__ ((interrupt)) void keyboard_handler(t_int_frame *int_frame)
+{
+	kprintf("KEYBOARD IRQ\n");
+	LOAD_INTERRUPT_STACK;
+	pic_8259_eoi(1);
+	RESET_INTERRUPT_STACK;
+}
```

Comment:
I think it's better to not doing anythink outside LOAD/RESET_INTERRUPT_STACK interval.

Reply:
Ok but we must keep the EOI it's a requirement after an IRQ is fired by the PIC.

Reply:
The frame we want to print is the caller frame, NOT the one we setup. So we should obviously print the frame BEFORE the `LOAD_INTERRUPT_STACK`.

In the same way, the `pic_8259_eoi` call will enable the PIC again to receive interrupts, so I think it's the last thing to do in our handler (after the `RESET_INTERRUPT_STACK`).

Those LOAD/RESET macros only resets stack segments registers, which are not used by regular functions (fix me if I'm wrong). The only thing that stores/restores the general purpose registers and thus must be called at the very top & at the very and are the `pusha`/`popa`/`iret` instructions, and this is handled well thanks to the interrupt attribute.

Reply:
The Marco LOAD/RESET_INTERRUPT_STACK is not named correctly. In reality it's loading the kernel data segment. And the frame of the caller is push to the stack by the cpu (first argument of the function) so it's preferable to load the kernel data segment at first doing think to handle the interrupt and at the end reset the the previous segment.